File: interaccion/src/dialogo_nodo.cpp
//Se introducen las cabeceras
#include "ros/ros.h"
#include "interaccion/usuario.h"
#include "interaccion/multiplicador.h"
#include "std_msgs/String.h"
#include "std_msgs/Bool.h"
#include <string>

/**
 * Se implementa un nodo que espera recibir mensajes cuyo topic es "user_topic" del tipo interaccion::usuario.
 * Muestra en pantalla este mensaje recibido
 */

//Muestra que le han llegado todos los datos
bool confirmarStart = false;
bool confirmarReset = false;

//Se pone a true tras enviar el primer mensaje
bool starting = false;

//Declaracion de las variables de comando de voz
std::string text;
std::string command;

//Se crean srv y client como variables globales
interaccion::multiplicador srv;
ros::ServiceClient client;

//Informacion a sintetizar por voz
std::string name;
std::string age;
std::string emotion;
std::string xposition;
std::string yposition;
std::string zposition;
std::string language;

/**
* Esta función muestra por pantalla el mensaje recibido desde el nodo empaquetador
*/
void functionCallback(const interaccion::usuario::ConstPtr& msg){
 ROS_INFO("El nombre es: %s", msg->infPersonal.nombre.c_str());;
 ROS_INFO("La edad es: %d", msg->infPersonal.edad);
 for (int i = 0; i < msg->infPersonal.idiomas.size(); i++) {
   ROS_INFO("Los idiomas que se son: %s", msg->infPersonal.idiomas[i].c_str());
 }
 ROS_INFO("La emocion es: %s", msg->emocion.c_str());
 ROS_INFO("La posicion en x es: %d", msg->posicion.x);
 ROS_INFO("La posicion en y es: %d", msg->posicion.y);
 ROS_INFO("La posicion en z es: %d", msg->posicion.z);

 srv.request.entrada = msg->infPersonal.edad; //Se quiere multiplicar la edad
 if (client.call(srv)){
   ROS_INFO("La edad multiplicada por dos es: %d", (int)srv.response.resultado);
 }else{
   ROS_ERROR("Fallo al llamar al servicio: multiplicador.srv");
 }
 
 //Sintetizador de voz
 name = msg->infPersonal.nombre.c_str();
 age = std::to_string(msg->infPersonal.edad);
 emotion = msg->emocion.c_str();
 xposition = std::to_string(msg->posicion.x);
 yposition = std::to_string(msg->posicion.y);
 zposition = std::to_string(msg->posicion.z);
 
 //Los datos son sintetizados por voz de manera ordenada
 command = "espeak -v es \"" "Mi nombre es "+name+"y tengo"+age+"anyos. Mi estado de animo es"+emotion+" \"";
 system(command.c_str());
 
 command = "espeak -v es \"" "Mi posicion en coordenadas cartesinas es "+xposition+" "+yposition+" "+zposition+" \"";
 system(command.c_str());

 command = "espeak -v es \"" "Además, hablo los siguientes idiomas" " \"";
 system(command.c_str());

 for (int i = 0; i < msg->infPersonal.idiomas.size(); i++) {
   language = msg->infPersonal.idiomas[i].c_str();
   command = "espeak -v es \"" +language+ " \"";
   system(command.c_str());
 }
 
 /*Cuando arranca el nodo, se confirma el start, a partir de ahí, se envía Reset
 info para nodo reloj*/
 if (starting == false) {
 	confirmarStart = true;
 } else {
 	confirmarReset = true;
 }
}

/*Funcion que recibe la informacion still_alive del nodo reloj
Si true, el nodo sigue vivo*/

void functionCallback2(const std_msgs::Bool::ConstPtr& msg){
	ROS_INFO("El estado del nodo reloj es: %s", msg ? "true" : "false");
}

//Función principal
int main(int argc, char **argv){
 //registra el nombre del nodo: dialogo_nodo y se crea el handle, denominado nodoDialogo
 ros::init(argc, argv, "dialogo_nodo");
 ros::NodeHandle nodoDialogo;
 ROS_INFO("dialogo_nodo creado y registrado"); //muestra en pantalla

 //si se recibe el mensaje cuyo topic es: "user_topic" se llama a la función manejadora: functionCallback
 ros::Subscriber subscriptorDialogo = nodoDialogo.subscribe("user_topic", 0, functionCallback);
 
 //si se recibe el mensaje cuyo topic es: "still_alive" se llama a la función manejadora: funcionCallback2
 ros::Subscriber subscriptorReloj = nodoDialogo.subscribe("still_alive", 0, functionCallback2);

  //Se invoca el servicio llamado Multiplicador
 client = nodoDialogo.serviceClient<interaccion::multiplicador>("multiplicador");

 //Se publica el mensaje start
 ros::Publisher relojStart = nodoDialogo.advertise<std_msgs::String>("start_topic",0);

  //Se publica el mensaje reset
 ros::Publisher relojReset = nodoDialogo.advertise<std_msgs::String>("reset_topic",0);
 
 while (ros::ok()){
 	 //Cuando el nodo se ejecuta por primera vez, publica el mensaje start
	 if (confirmarStart == true){
	 	std_msgs::String mensajeStart;
	 	std::string start = "start";
	 	mensajeStart.data = start;

	 	relojStart.publish(mensajeStart);

	 	//Una vez publicado, starting = true, indicando que a partir de ahora, se envía reset
	 	confirmarStart = false;
	 	starting = true;
	 }

	 //A partir de que el nodo se ha ejecutado por primera vez, publica el mensaje reset
 	 if (confirmarReset == true){
 		std_msgs::String mensajeReset;
 		std::string reset = "reset";
 		mensajeReset.data = reset;

 		relojReset.publish(mensajeReset);

  		confirmarReset = false;
	 }
	 ros::spinOnce();
	}

 return 0;
}

File: interaccion/src/dialogo_nodo.py
#SE IMPORTA ROSPY, LOS MENSAJES Y LAS LIBRERIAS QUE SE VAN A USAR
import rospy
from trabajo.msg import usuario,testAction,testGoal
from std_msgs.msg import Bool,String
from trabajo.srv import duplica
import os
import actionlib
#FUNCION QUE DEFINE EL CLIENTE DEL ACTIONLIB
def action_lib_client(name):
	#SE DEFINE EL CLIENTE
	client = actionlib.SimpleActionClient('deletrea', testAction)

	#SE COMPRUEBA SI EL CLIENTE RESPONDE EN UN PLAZO CORTO (0.5S)
	if client.wait_for_server(rospy.Duration(0.5)):

		#DEFINIMOS EL GOAL, EN ESTE CASO EL NOMBRE A DELETREAR
		goal = testGoal(name)

		#LO ENVIA
		client.send_goal(goal)

		#SE ESPERA QUE ACABE DE DELETREAR PARA CONTINUAR LA EJECUCION
		client.wait_for_result()

	#SI NO TIENE RESPUESTA DEL SERVIDOR INDICA QUE ESTA APAGADO
	else:
		print("El actionlib no esta conectado")


#CLIENTE DEL SERVICIO MATEMATICO
def duplica_cliente(edad):
	#INTENTA HACER LA PETICION AL SERVICIO, INDICANDO QUE NO ESTA CONECTADO
	#EN CASO DE APARECER LA EXCEPTION
    try:
		#SE DEFINE EL CLIENTE
        dup = rospy.ServiceProxy('duplica', duplica)

		#SE HACE LA PETICION CON LA EDAD
        resp = dup(edad)

		#SE IMPRIME POR PANTALLA LA EDAD
        rospy.loginfo("El doble de la edad es: %d", resp.edadx2)


    except rospy.ServiceException as e:
        print("El nodo matematico no esta conectado")

#CALLBACK EN CASO DE RECIBIR UN MENSAJE DEL EMPAQUETADOR
def callback(data):
	#DEFINO COMO GLOBAL ALGUNA VARIABLE QUE USARÉ
    global first,start,reset

	#COMPRUEBA SI ES EL PRIMER MENSAJE QUE RECIBE, EN ESE CASO LO PUBLICA AL RELOJ
	#POR EL PUBLICADOR DE START
    if first:
        start.publish("Start")

		#SE PONE LA FLAG A FALSE PARA QUE NO VUELVA A ENTRAR
        first=False

	#LAS SIGUIENTES VECES LO HARÁ POR EL DE RESET
    else:
        reset.publish("Reset")

	#IMPRIME EL NOMBRE RECIBIDO
    rospy.loginfo("Nombre: %s" ,data.inf.nombre)

	#CREA LA CADENA DE CARÁCTERES DEL COMANDO
    scommand='espeak -ves+m3' +' "'+"Se llama: " +data.inf.nombre + '"'

	#HACE LA LLAMADA AL SISTEMA CON EL COMANDO A LANZAR, EN ESTE CASO EL SINTETIZADOR
    os.system(scommand)

	#HACE UNA PETICION AL ACTIONLIB SERVER
    action_lib_client(data.inf.nombre)

    #IMPRIME LA EDAD
    rospy.loginfo("Edad: %d" ,data.inf.edad)

	#CREA LA CADENA DE CARÁCTERES DEL COMANDO
    scommand='espeak -ves+m3' +' "'+"Tiene: " +str(data.inf.edad) + " años"+'"'

	#HACE LA LLAMADA AL SISTEMA CON EL COMANDO A LANZAR, EN ESTE CASO EL SINTETIZADOR
    os.system(scommand)

	#HACE LA LLAMADA AL SERVICIO MATEMATICO
    duplica_cliente(data.inf.edad)

	#IMPRIME LOS IDIOMAS QUE HABLA
    rospy.loginfo("Idiomas: %s" ,data.inf.idiomas)

	#CREA CON LOS IDIOMAS UNA CADENA DE TEXTO LEGIBLE, ESTO ES PARA QUE AL SINTETIZAR
	#EL TEXTO QUEDE MAS NATURAL, AÑADIENDO LA CONJUNCION "Y" ENTRE LOS DOS ULTIMOS
	#IDIOMAS DE LA LISTA Y COMAS ENTRE EL RESTO
    idioma_str=data.inf.idiomas[0]
    if len(data.inf.idiomas)>1:
        for i in range(1,len(data.inf.idiomas)):
            if i==len(data.inf.idiomas)-1:
                idioma_str=idioma_str + " y "
            else:
                idioma_str=idioma_str+ ", "
            idioma_str=idioma_str+data.inf.idiomas[i]
    #SE CREA LA CADENA COMPLETA DEL COMANDO
    scommand='espeak -ves+m3' +' "'+"Habla: " +idioma_str + '"'

	#SE LANZA EL COMANDO
    os.system(scommand)

	#SE IPRIME LA EMOCION
    rospy.loginfo("Emocion: %s",data.emocion)

	#SE CREA LA CADENA DE TEXTO DEL COMANDO CON LA EMOCION
    scommand='espeak -ves+m3' +' "'+"Se encuentra: " +data.emocion + '"'

	#SE LANZA EL COMANDO
    os.system(scommand)

	#FINALMENTE SE IMPRIME LA POSICION, EN ESTE CASO NO SE SINTETIZA
    rospy.loginfo("Posicion: X:%d    Y:%d    Z:%d",data.posicion.x ,data.posicion.y ,data.posicion.z)


#CALLBACK DEL MENSAJE DE STILL_ALIVE
def still_alive(data):
	#LO INDICA POR PANTALLA
    rospy.loginfo("--------EL NODO RELOJ SIGUE VIVO--------")

#FUNCION PRINCIPAL
def dialogo():
	#DEFINIMOS LAS VARIABLES QUE SE USARAN COMO GLOBALES
    global first,start,reset

	#SE INICIALIZA A TRUE EL FLAG DE PRIMER MENSAJE
    first=True

	#SE INICIA EL NODO
    rospy.init_node('dialogo', anonymous=True)

	#SE DEFINE LOS PUBLICADORES
    start=rospy.Publisher('start_topic', String, queue_size=1)
    reset=rospy.Publisher('reset_topic', String, queue_size=1)

	#SE DEFINEN LOS SUBSCRIPTORES
    rospy.Subscriber('user_topic', usuario, callback)
    rospy.Subscriber('still_alive', Bool, still_alive)

	#SE PONE A CORRER
    rospy.spin()

if __name__=="__main__":
	#INICIAMOS
    dialogo()
